src/hotspot/share/opto/macro.cpp

```diff
@@ -1810,74 +1798,93 @@ void PhaseMacroExpand::jfr_sample_fast_object_allocation(
-    Node* alloc_count = make_load(alloc_sample_enabled_ctrl, alloc_sample_enabled_mem, tls, alloc_count_offset, TypeLong::LONG, T_LONG);
-    Node* alloc_count_new = transform_later(new AddLNode(alloc_count, longcon(1)));
-    alloc_sample_enabled_mem = make_store(alloc_sample_enabled_ctrl, alloc_sample_enabled_mem, tls, alloc_count_offset, alloc_count_new, T_LONG);
-    const int alloc_count_until_sample_offset = in_bytes(TRACE_THREAD_ALLOC_COUNT_UNTIL_SAMPLE_OFFSET);
-    Node* alloc_count_until_sample = make_load(alloc_sample_enabled_ctrl, alloc_sample_enabled_mem, tls, alloc_count_until_sample_offset, TypeLong::LONG, T_LONG);
-    Node* alloc_count_until_sample_cmp = transform_later(new CmpLNode(alloc_count_until_sample, alloc_count_new));
-    Node* alloc_sample_hit_bool = transform_later(new BoolNode(alloc_count_until_sample_cmp, BoolTest::eq));
-    IfNode* alloc_sample_hit_if = (IfNode*)transform_later(new IfNode(alloc_sample_enabled_ctrl, alloc_sample_hit_bool, PROB_MIN, COUNT_UNKNOWN));
-    Node* alloc_sample_hit_ctrl = transform_later(new IfTrueNode(alloc_sample_hit_if));
-    Node* alloc_sample_hit_mem = alloc_sample_enabled_mem;
-    Node* alloc_sample_miss_ctrl = transform_later(new IfFalseNode(alloc_sample_hit_if));
-    Node* alloc_sample_miss_mem = alloc_sample_enabled_mem;
-    Node* alloc_sample_hit_region = transform_later(new RegionNode(3));
-    Node* alloc_sample_hit_region_phi_mem = transform_later(new PhiNode(alloc_sample_hit_region, Type::MEMORY, TypeRawPtr::BOTTOM));
-
-    // if sample_hit then
-    {
-      CallLeafNode *call = new CallLeafNode(OptoRuntime::jfr_fast_object_alloc_Type(),
-                                            CAST_FROM_FN_PTR(address, OptoRuntime::jfr_fast_object_alloc_C),
-                                            "jfr_fast_object_alloc_C",
-                                            TypeRawPtr::BOTTOM);
-      call->init_req(TypeFunc::Parms+0, fast_oop);
-      call->init_req(TypeFunc::Parms+1, intcon(bottom_java_frame_bci(alloc->jvms())));
-      call->init_req(TypeFunc::Parms+2, tls);
-      call->init_req(TypeFunc::Control, alloc_sample_hit_ctrl);
-      call->init_req(TypeFunc::I_O    , top());
-      call->init_req(TypeFunc::Memory , alloc_sample_hit_mem);
-      call->init_req(TypeFunc::ReturnAdr, alloc->in(TypeFunc::ReturnAdr));
-      call->init_req(TypeFunc::FramePtr, alloc->in(TypeFunc::FramePtr));
-      transform_later(call);
-      alloc_sample_hit_ctrl = new ProjNode(call,TypeFunc::Control);
-      transform_later(alloc_sample_hit_ctrl);
-      alloc_sample_hit_mem = new ProjNode(call,TypeFunc::Memory);
-      transform_later(alloc_sample_hit_mem);
-
-      alloc_sample_hit_region->init_req(enabled_idx, alloc_sample_hit_ctrl);
-      alloc_sample_hit_region_phi_mem->init_req(enabled_idx, alloc_sample_hit_mem);
+     JVMState* old_state = alloc->jvms();
+     JVMState* new_state = old_state->clone_deep(C);
```

Comment:
Maybe we could simply extract this part of the code to another function? It seems nearly the same as `CallJavaNode::copy_call_debug_info`.

Reply:
Also, this code may need testing. Seems we need to add tests to verify this quite big change.

Reply:
BTW, I think maybe we need someone like Teacher Kuai to have some reviews.

Reply:
I just found out that `copy_call_debug_info` of 11 will deep clone the JVM states, so use this method to replace these operations here:)